File: backend/src/integrations/kms/pkcs11_hsm_connector.js
/**
 * PKCS#11 / Hardware Security Module (HSM) Connector — Phase 14.3
 *
 * Implements metadata collection for PKCS#11 compliant Hardware Security Modules
 * (e.g., Thales Luna, Utimaco, AWS CloudHSM, Nitrokey, SoftHSM).
 *
 * CRITICAL SECURITY INVARIANTS:
 * 1. Read-Only Token Session: Opened with CKF_SERIAL_SESSION (never CKF_RW_SESSION).
 * 2. Never Query CKA_VALUE: CKA_VALUE for private/secret keys is strictly excluded from attribute templates.
 * 3. Only Public & Operational Attributes: CKA_CLASS, CKA_KEY_TYPE, CKA_LABEL, CKA_ID,
 *    CKA_MODULUS_BITS, CKA_VALUE_LEN, CKA_EXTRACTABLE, CKA_NEVER_EXTRACTABLE, CKA_ENCRYPT, CKA_DECRYPT, etc.
 */

const { BaseKmsConnector } = require("./base_kms_connector");
const { KmsKeyMetadata, ProtectedKeyMaterialError } = require("./kms_metadata");

// PKCS#11 Attribute Constants (standard RFC 7512 / PKCS#11 v2.40/v3.0)
const CKA_ALLOWED_ATTRIBUTES = Object.freeze([
  "CKA_CLASS",
  "CKA_KEY_TYPE",
  "CKA_LABEL",
  "CKA_ID",
  "CKA_MODULUS_BITS",
  "CKA_VALUE_LEN",
  "CKA_EXTRACTABLE",
  "CKA_NEVER_EXTRACTABLE",
  "CKA_SENSITIVE",
  "CKA_ALWAYS_SENSITIVE",
  "CKA_ENCRYPT",
  "CKA_DECRYPT",
  "CKA_SIGN",
  "CKA_VERIFY",
  "CKA_WRAP",
  "CKA_UNWRAP",
  "CKA_TOKEN",
]);

const CKA_FORBIDDEN_ATTRIBUTES = Object.freeze([
  "CKA_VALUE", // Raw key bytes
  "CKA_PRIVATE_EXPONENT", // RSA d
  "CKA_PRIME_1", // RSA p
  "CKA_PRIME_2", // RSA q
  "CKA_EXPONENT_1", // RSA d mod (p-1)
  "CKA_EXPONENT_2", // RSA d mod (q-1)
  "CKA_COEFFICIENT", // RSA q^-1 mod p
  "CKA_VALUE_BITS",
]);

class Pkcs11HsmConnector extends BaseKmsConnector {
  /**
   * @param {Object} options
   * @param {string} options.name
   * @param {Object} [options.config]
   * @param {string} [options.config.modulePath] - Path to PKCS#11 library (.so / .dll / .dylib)
   * @param {number|string} [options.config.slotId=0] - Token slot ID
   * @param {string} [options.config.tokenLabel] - Label of token
   * @param {string} [options.config.userPin] - User PIN for read-only session
   * @param {Object} [options.client] - PKCS#11 mock or library wrapper
   * @param {Function} [options.fetchFn]
   */
  constructor(options = {}) {
    super({
      ...options,
      provider: "pkcs11_hsm",
    });
  }

  validateConfig(config) {
    super.validateConfig(config);
    if (config.slotId === undefined) {
      this.config = { ...config, slotId: 0 };
    }
  }

  getLeastPrivilegeRoleDefinition() {
    return {
      sessionType: "CKF_SERIAL_SESSION",
      accessType: "ReadOnly (User/Public session)",
      allowedAttributes: [...CKA_ALLOWED_ATTRIBUTES],
      forbiddenAttributes: [...CKA_FORBIDDEN_ATTRIBUTES],
      description: "Least-privilege read-only PKCS#11 token session querying public descriptors only.",
    };
  }

  validateLeastPrivilege(requestedAttributes = []) {
    const list = Array.isArray(requestedAttributes) ? requestedAttributes : [];
    const violations = [];

    for (const attr of list) {
      const upper = String(attr).toUpperCase();
      if (CKA_FORBIDDEN_ATTRIBUTES.includes(upper)) {
        violations.push(`Violation: Forbidden attribute '${attr}' exposes protected cryptographic material.`);
      }
    }

    return {
      valid: violations.length === 0,
      violations,
    };
  }

  async testConnection() {
    if (this.client && typeof this.client.C_GetTokenInfo === "function") {
      try {
        const info = await this.client.C_GetTokenInfo(this.config.slotId || 0);
        return { ok: true, message: `PKCS#11 token connected: ${info.label || "HSM"}` };
      } catch (err) {
        return { ok: false, message: `PKCS#11 token probe failed: ${err.message}` };
      }
    }
    return { ok: true, message: `PKCS#11 HSM connector configured for slot ${this.config.slotId}` };
  }

  async listKeys() {
    if (this.client && typeof this.client.findObjects === "function") {
      const objects = await this.client.findObjects({
        slotId: this.config.slotId,
        attributes: [
          { type: "CKA_CLASS", values: ["CKO_PRIVATE_KEY", "CKO_SECRET_KEY", "CKO_PUBLIC_KEY"] },
        ],
      });
      return objects.map((o) => o.id || o.label || `hsm-obj-${o.handle}`);
    }
    return this.config.mockKeys || [];
  }

  async describeKey(keyIdentifier) {
    let rawObj = null;

    if (this.client && typeof this.client.getAttributeValues === "function") {
      // Strictly query safe attributes only!
      const attrs = await this.client.getAttributeValues(keyIdentifier, CKA_ALLOWED_ATTRIBUTES);
      rawObj = attrs;
    } else if (this.config.mockKeyDetails && this.config.mockKeyDetails[keyIdentifier]) {
      rawObj = this.config.mockKeyDetails[keyIdentifier];
    } else {
      throw new Error(`PKCS#11 HSM: Key object '${keyIdentifier}' not found`);
    }

    // Explicit security check
    this.assertNoHsmPrivateMaterial(rawObj);

    return this.mapToKmsMetadata(keyIdentifier, rawObj);
  }

  assertNoHsmPrivateMaterial(obj) {
    if (!obj || typeof obj !== "object") return;
    for (const forbidden of CKA_FORBIDDEN_ATTRIBUTES) {
      if (forbidden in obj || forbidden.toLowerCase() in obj) {
        throw new ProtectedKeyMaterialError(
          `CRITICAL SECURITY VIOLATION: HSM attribute '${forbidden}' detected. ECDAT never extracts private key material!`
        );
      }
    }
  }

  mapToKmsMetadata(keyIdentifier, rawObj) {
    const keyType = (rawObj.CKA_KEY_TYPE || rawObj.keyType || "CKK_RSA").toUpperCase();
    const modulusBits = rawObj.CKA_MODULUS_BITS || rawObj.modulusBits;
    const valueLen = rawObj.CKA_VALUE_LEN || rawObj.valueLen;

    const { algorithm, size } = this.parseHsmAlgorithm(keyType, modulusBits, valueLen);

    const label = rawObj.CKA_LABEL || rawObj.label || keyIdentifier;
    const id = rawObj.CKA_ID || rawObj.id || keyIdentifier;
    const slotId = this.config.slotId !== undefined ? this.config.slotId : 0;
    const canonicalKeyId = `pkcs11:slot=${slotId};id=${encodeURIComponent(id)};label=${encodeURIComponent(label)}`;

    const owner = this.config.tokenLabel || `slot-${slotId}`;

    const isExtractable = Boolean(rawObj.CKA_EXTRACTABLE ?? rawObj.extractable ?? false);
    const neverExtractable = Boolean(rawObj.CKA_NEVER_EXTRACTABLE ?? rawObj.neverExtractable ?? true);

    const operations = [];
    if (rawObj.CKA_ENCRYPT || rawObj.CKA_DECRYPT) operations.push("encrypt", "decrypt");
    if (rawObj.CKA_SIGN || rawObj.CKA_VERIFY) operations.push("sign", "verify");
    if (rawObj.CKA_WRAP || rawObj.CKA_UNWRAP) operations.push("wrapKey", "unwrapKey");

    const usage = {
      keyUsage: operations.includes("sign") ? "SIGN_VERIFY" : "ENCRYPT_DECRYPT",
      operations: operations.length > 0 ? operations : ["encrypt", "decrypt"],
      origin: "Hardware Security Module (HSM)",
      isExportable: isExtractable && !neverExtractable,
    };

    const rotation = {
      enabled: false, // HSM keys typically rely on external lifecycle management
      periodDays: null,
      lastRotatedAt: null,
      nextRotationAt: null,
      version: "1",
    };

    return new KmsKeyMetadata({
      keyId: canonicalKeyId,
      algorithm,
      size,
      state: "Active",
      rotation,
      owner,
      usage,
      provider: "pkcs11_hsm",
      description: `Hardware Security Module Token Key: ${label}`,
      rawMetadata: {
        slotId,
        label,
        id,
        keyType,
        modulusBits,
        extractable: isExtractable,
        neverExtractable,
        token: rawObj.CKA_TOKEN ?? true,
      },
    });
  }

  parseHsmAlgorithm(keyType, modulusBits, valueLen) {
    const kt = String(keyType).toUpperCase();
    if (kt.includes("RSA")) {
      const bits = modulusBits || 2048;
      return { algorithm: `RSA-${bits}`, size: bits };
    }
    if (kt.includes("EC") || kt.includes("ECDSA")) {
      return { algorithm: "ECDSA-P256", size: 256 };
    }
    if (kt.includes("AES")) {
      const bits = (valueLen || 32) * 8;
      return { algorithm: `AES-${bits}-GCM`, size: bits };
    }
    if (kt.includes("DES3") || kt.includes("3DES")) {
      return { algorithm: "3DES-EDE", size: 168 };
    }
    return { algorithm: kt, size: modulusBits || 256 };
  }
}

module.exports = {
  Pkcs11HsmConnector,
  CKA_ALLOWED_ATTRIBUTES,
  CKA_FORBIDDEN_ATTRIBUTES,
};

File: scanners/integrations/kms/pkcs11_hsm_connector.py
"""
PKCS#11 / Hardware Security Module (HSM) Connector (Python) — Phase 14.3

Implements metadata collection for PKCS#11 compliant HSMs.
CRITICAL INVARIANT: Never query or extract CKA_VALUE or private key material.
"""

from typing import Dict, Any, List, Optional
import urllib.parse
from .base_kms_connector import BaseKmsConnector
from .kms_metadata import KmsKeyMetadata, ProtectedKeyMaterialError

CKA_ALLOWED_ATTRIBUTES = [
    "CKA_CLASS",
    "CKA_KEY_TYPE",
    "CKA_LABEL",
    "CKA_ID",
    "CKA_MODULUS_BITS",
    "CKA_VALUE_LEN",
    "CKA_EXTRACTABLE",
    "CKA_NEVER_EXTRACTABLE",
    "CKA_SENSITIVE",
    "CKA_ALWAYS_SENSITIVE",
    "CKA_ENCRYPT",
    "CKA_DECRYPT",
    "CKA_SIGN",
    "CKA_VERIFY",
    "CKA_WRAP",
    "CKA_UNWRAP",
    "CKA_TOKEN",
]

CKA_FORBIDDEN_ATTRIBUTES = [
    "CKA_VALUE",
    "CKA_PRIVATE_EXPONENT",
    "CKA_PRIME_1",
    "CKA_PRIME_2",
    "CKA_EXPONENT_1",
    "CKA_EXPONENT_2",
    "CKA_COEFFICIENT",
    "CKA_VALUE_BITS",
]


class Pkcs11HsmConnector(BaseKmsConnector):
    def __init__(self, name: str, config: Optional[Dict[str, Any]] = None, client: Any = None):
        super().__init__(name=name, provider="pkcs11_hsm", config=config, read_only=True, client=client)

    def validate_config(self, config: Dict[str, Any]) -> None:
        super().validate_config(config)
        if "slot_id" not in config:
            self.config["slot_id"] = 0

    def get_least_privilege_role_definition(self) -> Dict[str, Any]:
        return {
            "session_type": "CKF_SERIAL_SESSION",
            "access_type": "ReadOnly (User/Public session)",
            "allowed_attributes": list(CKA_ALLOWED_ATTRIBUTES),
            "forbidden_attributes": list(CKA_FORBIDDEN_ATTRIBUTES),
            "description": "Least-privilege read-only PKCS#11 token session querying public descriptors only.",
        }

    def validate_least_privilege(self, attributes: Any) -> Dict[str, Any]:
        attr_list = attributes if isinstance(attributes, list) else []
        violations = []

        for a in attr_list:
            if str(a).upper() in CKA_FORBIDDEN_ATTRIBUTES:
                violations.append(f"Violation: Forbidden attribute '{a}' exposes protected cryptographic material.")

        return {"valid": len(violations) == 0, "violations": violations}

    def list_keys(self) -> List[str]:
        if self.client and hasattr(self.client, "find_objects"):
            objects = self.client.find_objects(slot_id=self.config.get("slot_id", 0))
            return [o.get("id") or o.get("label") for o in objects]
        return self.config.get("mock_keys", [])

    def describe_key(self, key_identifier: str) -> KmsKeyMetadata:
        raw_obj = None
        if self.client and hasattr(self.client, "get_attribute_values"):
            raw_obj = self.client.get_attribute_values(key_identifier, CKA_ALLOWED_ATTRIBUTES)
        elif "mock_key_details" in self.config and key_identifier in self.config["mock_key_details"]:
            raw_obj = self.config["mock_key_details"][key_identifier]
        else:
            raise ValueError(f"PKCS#11 HSM: Key object '{key_identifier}' not found")

        self._assert_no_hsm_private_material(raw_obj)
        return self._map_to_kms_metadata(key_identifier, raw_obj)

    def _assert_no_hsm_private_material(self, obj: Dict[str, Any]) -> None:
        if not isinstance(obj, dict):
            return
        for forbidden in CKA_FORBIDDEN_ATTRIBUTES:
            if forbidden in obj or forbidden.lower() in obj:
                raise ProtectedKeyMaterialError(
                    f"CRITICAL SECURITY VIOLATION: HSM attribute '{forbidden}' detected. ECDAT never extracts private key material!"
                )

    def _map_to_kms_metadata(self, key_identifier: str, raw_obj: Dict[str, Any]) -> KmsKeyMetadata:
        key_type = str(raw_obj.get("CKA_KEY_TYPE") or raw_obj.get("key_type") or "CKK_RSA").upper()
        modulus_bits = raw_obj.get("CKA_MODULUS_BITS") or raw_obj.get("modulus_bits")
        value_len = raw_obj.get("CKA_VALUE_LEN") or raw_obj.get("value_len")

        algorithm, size = self._parse_hsm_algorithm(key_type, modulus_bits, value_len)

        label = raw_obj.get("CKA_LABEL") or raw_obj.get("label") or key_identifier
        kid = raw_obj.get("CKA_ID") or raw_obj.get("id") or key_identifier
        slot_id = self.config.get("slot_id", 0)

        canonical_id = f"pkcs11:slot={slot_id};id={urllib.parse.quote(str(kid))};label={urllib.parse.quote(str(label))}"
        owner = self.config.get("token_label") or f"slot-{slot_id}"

        extractable = bool(raw_obj.get("CKA_EXTRACTABLE", raw_obj.get("extractable", False)))
        never_extractable = bool(raw_obj.get("CKA_NEVER_EXTRACTABLE", raw_obj.get("never_extractable", True)))

        ops = []
        if raw_obj.get("CKA_ENCRYPT") or raw_obj.get("CKA_DECRYPT"):
            ops.extend(["encrypt", "decrypt"])
        if raw_obj.get("CKA_SIGN") or raw_obj.get("CKA_VERIFY"):
            ops.extend(["sign", "verify"])
        if raw_obj.get("CKA_WRAP") or raw_obj.get("CKA_UNWRAP"):
            ops.extend(["wrapKey", "unwrapKey"])

        usage = {
            "key_usage": "SIGN_VERIFY" if "sign" in ops else "ENCRYPT_DECRYPT",
            "operations": ops or ["encrypt", "decrypt"],
            "origin": "Hardware Security Module (HSM)",
            "is_exportable": extractable and not never_extractable,
        }

        rotation = {
            "enabled": False,
            "period_days": None,
            "last_rotated_at": None,
            "next_rotation_at": None,
            "version": "1",
        }

        return KmsKeyMetadata(
            key_id=canonical_id,
            algorithm=algorithm,
            size=size,
            state="Active",
            owner=owner,
            rotation=rotation,
            usage=usage,
            provider="pkcs11_hsm",
            description=f"Hardware Security Module Token Key: {label}",
            raw_metadata={
                "slot_id": slot_id,
                "label": label,
                "id": kid,
                "key_type": key_type,
                "modulus_bits": modulus_bits,
                "extractable": extractable,
                "never_extractable": never_extractable,
                "token": raw_obj.get("CKA_TOKEN", True),
            },
        )

    def _parse_hsm_algorithm(self, key_type: str, modulus_bits: Optional[int], value_len: Optional[int]):
        kt = str(key_type).upper()
        if "RSA" in kt:
            bits = modulus_bits or 2048
            return f"RSA-{bits}", bits
        if "EC" in kt:
            return "ECDSA-P256", 256
        if "AES" in kt:
            bits = (value_len or 32) * 8
            return f"AES-{bits}-GCM", bits
        if "DES3" in kt or "3DES" in kt:
            return "3DES-EDE", 168
        return kt, modulus_bits or 256
